api.go

```diff
@@ -830,3 +834,112 @@ func (c Client) makeTargetURL(bucketName, objectName, bucketLocation string, que
+	if c.endpointURL.Scheme == "https" && strings.Contains(bucketName, ".") {
+		return false
+	}
+	return c.isVirtual
+}
+
+// probeVirtualHostStyle - sets the isVirtualHostStyle bool
+// to true if server can accept virtual hosted style urls.
+// Send a HEAD bucket request for a non existent bucket to
+// server with request url in virtual hosted style format,
+// and see if server responds with NotFound status.
+func (c *Client) probeVirtualHostStyle() {
+	// set default to false
+	c.isVirtual = false
+
+	host := c.endpointURL.Host
+	scheme := c.endpointURL.Scheme
+	method := "HEAD"
+	// Strip port 80 and 443 so we won't send these ports in Host header.
+	if h, p, err := net.SplitHostPort(host); err == nil {
+		if scheme == "http" && p == "80" || scheme == "https" && p == "443" {
+			host = h
+		}
+	}
+	// bucket doesn't exist.
+	bucketName := "auto-probe-bucketname"
+	// Construct HEAD bucket request with virtual style
+	urlStr := scheme + "://" + bucketName + "." + host + "/"
+
+	u, err := url.Parse(urlStr)
+	if err != nil {
+		return
+	}
+	// Initialize a new HTTP request for the method.
+	req, err := http.NewRequest(method, u.String(), nil)
+	if err != nil {
+		return
+	}
+
+	// Get credentials from the configured credentials provider.
+	value, err := c.credsProvider.Get()
+	if err != nil {
+		return
+	}
+
+	var (
+		signerType      = value.SignerType
+		accessKeyID     = value.AccessKeyID
+		secretAccessKey = value.SecretAccessKey
+		sessionToken    = value.SessionToken
+	)
+
+	// Custom signer set then override the behavior.
+	if c.overrideSignerType != credentials.SignatureDefault {
+		signerType = c.overrideSignerType
+	}
+
+	// If signerType returned by credentials helper is anonymous,
+	// then do not sign regardless of signerType override.
+	if value.SignerType == credentials.SignatureAnonymous {
+		signerType = credentials.SignatureAnonymous
+	}
+	// Set 'User-Agent' header for the request.
+	c.setUserAgent(req)
+
+	metadata := requestMetadata{
+		bucketName:       bucketName,
+		contentSHA256Hex: emptySHA256Hex,
+	}
+	// Set all headers.
+	for k, v := range metadata.customHeader {
+		req.Header.Set(k, v[0])
+	}
+
+	req.Body = nil
+	// Set incoming content-length.
+	req.ContentLength = metadata.contentLength
+	switch {
+	case signerType.IsV2():
+		// Add signature version '2' authorization header.
+		req = s3signer.SignV2(*req, accessKeyID, secretAccessKey)
+	default:
+		// Set sha256 sum for signature calculation only with signature version '4'.
+		shaHeader := unsignedPayload
+		if metadata.contentSHA256Hex != "" {
+			shaHeader = metadata.contentSHA256Hex
+		}
+		req.Header.Set("X-Amz-Content-Sha256", shaHeader)
+		// Add signature version '4' authorization header.
+		req = s3signer.SignV4(*req, accessKeyID, secretAccessKey, sessionToken, "us-east-1")
+	}
+	// Initiate the request.
+	r, _ := c.do(req)
+	if r != nil && (r.StatusCode == http.StatusNotFound || r.StatusCode == http.StatusOK) {
```

Comment:
A more succinct way of expressing this would be,
```go
switch r.StatusCode {
case http.StatusNotFound, http.StatusOK:
  c.isVirtual = true
}
```

Reply:
There is a still a problem here:
If a random user creates 'auto-probe-bucketname' in AWS S3, all other users will get access denied after a HEAD, which means that they will never have virtual host enabled.

https://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketHEAD.html

```
13:14 $ aws s3api head-bucket --bucket aws

An error occurred (403) when calling the HeadBucket operation: Forbidden
```

(aws is a bucket created by another user)

Reply:
yes, @vadmeste - this is a little tricky. I have changed the check to using a random bucket name - this has a high probability of being non-existent. And even if it is, the client will switch to path style only for one invocation.